### ISSUE TITLE
Help: Additional information on flock

### ISSUE THREAD

**Issue by jaberbasma** [datalad-handbook/book#640]
Hi, 

We are trying to walk through the handbook and we are unfamiliar with flock, what it does, or what version is needed. 

Our cluster has flock but does not include the --verbose flag. We are using centos 7 on an HPC cluster. 

We are running into problems using flock and the script provided in the handbook and we are wondering if it could be due to version differences. 

Also please clarify the $DSLOCKFILE: 
Should it be an empty file or a directory? 
And should it be in the network file system shared version of the dataset or temp directory cloned on the compute node? 

Thank you, 
Basma

**Comment by adswa** [comment]
Hi  @jaberbasma thanks for opening the issue. I agree its a super dense section - feedback on what's unclear is very appreciated. I will try to briefly answer your questions in this issue but hope to improve the handbook section on this, too.

> Our cluster has flock but does not include the --verbose flag. We are using centos 7 on an HPC cluster.

The ``--verbose`` option leads to reports how long it takes to retrieve the lock. Its not fundamental for the functionality, but helpful info.

> We are running into problems using flock and the script provided in the handbook and we are wondering if it could be due to version differences.

I would be curious to hear about them. I hope that I or others can help. Would you mind describing your usecase and the problems here or in the datalad repo (whichever you prefer)?

> Also please clarify the $DSLOCKFILE:
Should it be an empty file or a directory?
And should it be in the network file system shared version of the dataset or temp directory cloned on the compute node?

In the examples in the handbook, we define $DSLOCKFILE as ``ENV(PWD)/.git/datalad_lock`` (which is an empty textfile in the .git/ directory of the original dataset, not in temp). flock will create this file if it doesn't exist. The purpose of flock and this file is the following:

Here's a late night attempt at explaining what flock does in our examples:

In case you have several parallel jobs that create new branches in their temporary clones and push them back once they are done, it can happen that one job is currently pushing back its result branch, while another jobs attempts to clone the dataset that is just in this very moment receiving a new branch, or yet another job also tries to push its branch. Such concurrency events get more likely the more jobs run in parallel, and if they happen, they can cause the clone or push operation that intervened with another push to fail or misbehave. Therefore, we use flock to safeguard these situations and prevent them from happening. Essentially, we make sure that a call is only executed when flock has made sure that no other process is modifying the original dataset.

In general, flock creates a file (the datalad_lock file, but it can be any file) and locks it until the command after flock has completed. Only if it can get this lock, it executes the command that is specified afterwards. In the examples in the handbook, all parallel runs share a single lockfile. If flock A it can't get the lock, it means that another job currently has is, and flock A waits with command execution until it gets the lock (typically a few milliseconds to seconds). When we use flock in front of any command that modifies the original dataset, we make sure that none of these processes run at the exact same time, because they essentially wait for eachother to finish first.

**Comment by adswa** [comment]
FTR: We are currently piloting a workflow with very reduced use of flock. It will take some time until its written up, but I can share code here if you are interested

**Comment by adswa** [comment]
Hey @jaberbasma , is this still relevant? I will close this due to inactivity, but please reopen or open a new issue if there are remaining questions. Also, we have created an updated workflow with a tutorial that is hopefully helpful. It can currently be found at https://github.com/psychoinformatics-de/processing-workflow (we may change the repo name in a bit, but not the organization)
